pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -92,72 +85,45 @@ func getRevisionLimits(operatorSpec *operatorv1.StaticPodOperatorSpec) (int32, i
 	if operatorSpec.SucceededRevisionLimit != 0 {
 		succeededRevisionLimit = operatorSpec.SucceededRevisionLimit
 	}
-	return failedRevisionLimit, succeededRevisionLimit
+	return int(failedRevisionLimit), int(succeededRevisionLimit)
 }
 
-func (c *PruneController) excludedRevisionHistory(ctx context.Context, recorder events.Recorder, failedRevisionLimit, succeededRevisionLimit, abandonedRevisionLimit int32) ([]int, error) {
-	var succeededRevisions, failedRevisions, inProgressRevisions, unknownStatusRevisions, abandonedRevisions []int
-
-	configMaps, err := c.configMapGetter.ConfigMaps(c.targetNamespace).List(ctx, metav1.ListOptions{})
-	if err != nil {
-		return []int{}, err
-	}
-	for _, configMap := range configMaps.Items {
-		if !strings.HasPrefix(configMap.Name, statusConfigMapName) {
-			continue
+func (c *PruneController) excludedRevisions(status *operatorv1.StaticPodOperatorStatus, failedRevisionLimit, succeededRevisionLimit int) sets.Int32 {
+	keep := sets.Int32{}
+	keep.Insert(int32RangeBelowOrEqual(status.LatestAvailableRevision, failedRevisionLimit)...)
+	for _, ns := range status.NodeStatuses {
+		if ns.CurrentRevision != 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.CurrentRevision, succeededRevisionLimit)...)
 		}
-
-		if revision, ok := configMap.Data["revision"]; ok {
-			revisionNumber, err := strconv.Atoi(revision)
-			if err != nil {
-				return []int{}, err
-			}
-			switch configMap.Data["status"] {
-			case string(corev1.PodSucceeded):
-				succeededRevisions = append(succeededRevisions, revisionNumber)
-			case string(corev1.PodFailed):
-				failedRevisions = append(failedRevisions, revisionNumber)
-			case StatusAbandoned:
-				abandonedRevisions = append(abandonedRevisions, revisionNumber)
-			case StatusInProgress:
-				// we always protect inprogress
-				inProgressRevisions = append(inProgressRevisions, revisionNumber)
-
-			default:
-				// protect things you don't understand
-				unknownStatusRevisions = append(unknownStatusRevisions, revisionNumber)
-				recorder.Event("UnknownRevisionStatus", fmt.Sprintf("unknown status for revision %d: %v", revisionNumber, configMap.Data["status"]))
-			}
+		if ns.TargetRevision != 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.TargetRevision, failedRevisionLimit)...)
+		}
+		if ns.LastFailedRevision != 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.LastFailedRevision, failedRevisionLimit)...)
 		}
 	}
+	return keep
+}
```

Comment:
Comment about how this is an approximation of the previous (ideal?) behavior. 

I read this method as:

* Don't prune `LatestAvailableRevision` and the `failedRevisionLimit - 1` revisions before it.
* Don't prune a node's `CurrentRevision` and the `succeededRevisionLimit - 1` revisions before it.
* Don't prune a node's `TargetRevision` and the `failedRevisionLimit - 1` revisions before it.
* Don't prune a node's `LastFailedRevision` and the `failedRevisionLimit - 1` revisions before it.

At min, provide some insight into how the `failedRevisionLimit` is being used without actually examining a revisions actual status other than those implied by `LastFailedRevision`.

Reply:
:+1: added